### PR TITLE
internal: definition based hover functions

### PR DESCRIPTION
**Pull request by HKalbasi**
This is part of #10181 but since it is blocked and `hover.rs` is moving quickly so will cause conflicts, I submitted this PR.

This PR extract some parts of `hover` to `find_definition` (maybe this need to be moved to some other file?) and `hover_for_definition`, with those functions I will be able to calculate definition of every token, and calculate hover (and probably other queries) for each definition only once.

### CODE REVIEW

File: crates/ide/src/hover.rs [hunk 1]
```diff
@@ -117,6 +117,26 @@ pub(crate) fn hover(
...
+                    if let Some(hover) =
+                        hover_for_definition(&sema, file_id, def, &token.parent().unwrap(), config)
+                    {
+                        return Some(RangeInfo::new(token.text_range(), hover));
+                    }
+                }
+            }
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (HKalbasi) [reply crates/ide/src/hover.rs:139]: This section is just for attributes, so you mean macro calls in attributes or attributes in macro calls? Do you have a test for that?

[the remark this answers is not in the document]
Comment (HKalbasi) [reply crates/ide/src/hover.rs]: I fixed it and added a test.

By the way, maybe iterating on `descend_into_macros_many` should be moved to `find_definition` as well?

**Comment by HKalbasi** [comment]
I reused `find_definition` in go to, so I think it shouldn't be in `hover.rs`. where is the right place for it?

File: crates/ide/src/hover.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Veykril) [reply crates/ide/src/hover.rs]: I don't think adding that makes to `find_definition` makes sense as a lot of ide features want to descend and check for things more specifically/not even use `find_definition`.

**Comment by Veykril** [comment]
The right place for such a function is probably in `ide_db::defs` on `Definition` itself.

File: crates/ide/src/hover.rs [hunk 2]
```diff
@@ -117,7 +117,37 @@ pub(crate) fn hover(
...
+                            &token.parent().unwrap(),
+                            config,
+                        ) {
+                            return Some(RangeInfo::new(token.text_range(), hover));
+                        }
+                    }
+                }
+            }
+        }
+    }
```
Comment (Veykril) [inline crates/ide/src/hover.rs:148]: It's a bit unclear to me as to why this is being moved out of the main iteration

Comment (HKalbasi) [reply crates/ide/src/hover.rs]: Now derive token is in the find_definition (now Definition::from_node), and for lints it isn't a problem?

File: crates/ide/src/goto_definition.rs [hunk 3]
```diff
@@ -99,41 +80,31 @@ pub(crate) fn goto_definition(
...
-                full_range: TextRange::new(0.into(), size),
-                name: path.into(),
-                focus_range: None,
-                kind: None,
-                container_name: None,
-                description: None,
-                docs: None,
-            }])
-        }
-        None => try_resolve_derive_input_at(
```
Comment (Veykril) [inline crates/ide/src/goto_definition.rs:129]: Where did this arm go?

File: crates/ide/src/hover.rs (hunk 2 above)
Comment (Veykril) [reply crates/ide/src/hover.rs]: I still don't really understand why you are pulling that one out of all the other logic. But it's fine I'm gonna refactor this module in the coming week so I can address that.

**Comment by lnicola** [comment]
changelog internal (first contribution) refactor hover and go to definition lookups

**Pull request merged**